src/graph.py

```diff
@@ -752,6 +750,104 @@ def perform_GO_enrichment_on_clusters(nxgraph, clustering):
+    if mode == "membership": # Summarize single-run parameter search
+        dfs = [pd.Series(1, index=graphs["augmented_forest"].nodes(), name=paramstring) for paramstring, graphs in results.items()]
+    elif mode == "robustness": # Summarize randomized robustness
+        dfs = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["robustness"].rename(paramstring) for paramstring, graphs in results.items()]
+    elif mode == "specificity": # Summarize randomized specificity
+        dfs = [get_networkx_graph_as_dataframe_of_nodes(graphs["augmented_forest"])["specificity"].rename(paramstring) for paramstring, graphs in results.items()]
+    else:
+        logger.info("`mode` must be one of the following: 'membership', 'robustness', or 'specificity'.")
+        return
+    
+    df = pd.concat(dfs, axis=1).fillna(0)
+    
+    # df can get quite large with many sparse entries, so let's filter for the top_n entries
+    if top_n == -1: return df
+    
+    if len(df) > top_n: 
+        cutoff = sorted(df.sum(axis=1).tolist(), reverse=True)[top_n]
+        df = df[df.sum(axis=1) > cutoff]
+    
+    return df
+
+
+def get_robust_subgraph_from_randomizations(nxgraph, max_size=400, min_component_size=5): 
+    """
+    Given a graph with robustness attributes, take the top `max_size` robust nodes and
+    prune any "small" components. 
+
+    Arguments:
+        nxgraph (networkx.Graph): Network from randomization experiment
+        max_size (int): Max size of robust network
+
+    Returns: 
+        networkx.Graph: Robust network
+
+    TODO: Potential alternative approaches
+    Approach 1: from entire network, attempt to remove lowest robustness node. If removal results in a component 
+    of size less than min_size, do not remove. 
+    Approach 2: select top max_size nodes based on robustness, then return subgraph. 
+    """
+    
+    if nxgraph.number_of_nodes() == 0: return nxgraph
+
+    node_attributes_df = get_networkx_graph_as_dataframe_of_nodes(nxgraph)
+
+    if "robustness" not in node_attributes_df.columns: 
+        logger.warning("'robustness' is not an attribute in subgraph.")
+        return None
+
+    # Get indices of top nodes sorted by high robustness, then low specificity. Don't allow nodes with robustness = 0.
+    node_attributes_df = node_attributes_df[node_attributes_df["robustness"] > 0]
+    node_attributes_df.sort_values(["robustness", "specificity"], ascending=[False, True], inplace=True)
+    top_hits = node_attributes_df.index[:min(max_size,len(node_attributes_df))]
+
+    # Get robust subnetwork and remove small components.
+    robust_network = nxgraph.subgraph(top_hits)
+    robust_network = filter_graph_by_component_size(robust_network, min_component_size)
+
+    return robust_network
+
+
+def filter_graph_by_component_size(nxgraph, min_size=5): 
+    # Removes any components that are less than `min_size`. Set `min_size`=2 to remove only singletons
+    
+    if nxgraph is None: return None
```

Comment:
Maybe add a log statement too? Also consider handling the case of empty networks, if that ever arises...